embed/trezorhal/image.c

```diff
@@ -29,27 +29,29 @@ static bool compute_pubkey(uint8_t sig_m, uint8_t sig_n, const uint8_t * const *
     return 0 == ed25519_cosi_combine_publickeys(res, keys, sig_m);
 }
 
-bool image_parse_header(const uint8_t *data, uint32_t magic, uint32_t maxsize, image_header *hdr)
+bool image_parse_header(const uint8_t * const data, const uint32_t magic, const uint32_t maxsize, image_header * const hdr)
 {
-    memcpy(&hdr->magic, data, 4);
+    uint32_t *data_ptr = (uint32_t *) data;
+
+    hdr->magic = *data_ptr++;
     if (hdr->magic != magic) return false;
 
-    memcpy(&hdr->hdrlen, data + 4, 4);
+    hdr->hdrlen = *data_ptr++;
     if (hdr->hdrlen != HEADER_SIZE) return false;
 
-    memcpy(&hdr->expiry, data + 8, 4);
-    if (hdr->expiry != 0) return false;
+    hdr->expiry = *data_ptr++;
+    if (hdr->expiry != 0) return false; // todo: then what is the point of having this field?
 
-    memcpy(&hdr->codelen, data + 12, 4);
-    if (hdr->hdrlen + hdr->codelen < 4 * 1024) return false;
-    if (hdr->hdrlen + hdr->codelen > maxsize) return false;
-    if ((hdr->hdrlen + hdr->codelen) % 512 != 0) return false;
+    hdr->codelen = *data_ptr++;
+    if ((hdr->hdrlen + hdr->codelen) < 4 * 1024) return false;
+    if ((hdr->hdrlen + hdr->codelen) > maxsize) return false;
+    if (((hdr->hdrlen + hdr->codelen) & 511) != 0) return false;
```

Comment:
I don't like this, it is much more cryptic and the original intention is more hidden. I was under impression that `% x` will get optimized to `& (x-1)` in case x is 2^n. If that's not the case, leave it, but in case it gets optimized, I strongly prefer keeping the original way.

Reply:
ok, will revert. same code is generated. this was just a habit of mine to avoid the remainder operator. it does not matter here.